File: Week1/henry/1-5.cpp
#include <string>
#include <vector>
#include <math.h>
using namespace std;
pair<int, int> nown(int a)
{
    if (a == 1)    return { 1,4 };
    if (a == 2)    return { 2,4 };
    if (a == 3)    return { 3,4 };
    if (a == 4)    return { 1,3 };
    if (a == 5)    return { 2,3 };
    if (a == 6)    return { 3,3 };
    if (a == 7)    return { 1,2 };
    if (a == 8)    return { 2,2 };
    if (a == 9)    return { 3,2 };
    if (a == 0)    return { 2,1 };
}
int dis(pair<int, int> a, pair<int, int> b) //거리 구하기
{
    return abs(a.first - b.first) + abs(a.second - b.second);
}
string solution(vector<int> numbers, string hand)
{
    string answer = "";
    int i;
    pair<int, int> L = { 1,1 }, R = { 3,1 }, n;
    for (i = 0; i < numbers.size(); ++i)
    {
        n = nown(numbers[i]);
        if (numbers[i] == 1 || numbers[i] == 4 || numbers[i] == 7) //예외
        {
            answer += 'L';
            L = n;
            continue;
        }
        if (numbers[i] == 3 || numbers[i] == 6 || numbers[i] == 9) //예외
        {
            answer += 'R';
            R = n;
            continue;
        }
        if (dis(n, L) > dis(n, R)) //2,5,8,0일 때 더 가까운 거리 구하기
        {
            answer += 'R';
            R = n;
        }
        else if (dis(n, L) < dis(n, R))
        {
            answer += 'L';
            L = n;
        }
        else //거리가 같으면 자기 메인 손으로 누르기
        {
            if (hand[0] == 'r')
            {
                answer += 'R';
                R = n;
            }
            else
            {
                answer += 'L';
                L = n;
            }
        }
    }
    return answer;
}

File: Week1/heystranger/1-5.py
def solution(numbers, hand):
    answer = ''

    # 좌표 설정
    position = {'1': (0, 0), '2': (0, 1), '3': (0, 2),
                '4': (1, 0), '5': (1, 1), '6': (1, 2),
                '7': (2, 0), '8': (2, 1), '9': (2, 2),
                '*': (3, 0), '0': (3, 1), '#': (3, 2)}

    # 오른손 왼손 위치
    right = position['#']
    left = position['*']

    for i in numbers:
        if i == 1 or i == 4 or i == 7:
            answer += 'L'
            left = position[str(i)]

        elif i == 3 or i == 6 or i == 9:
            answer += 'R'
            right = position[str(i)]

        else:
            l_distance = abs(left[0] - position[str(i)][0]) + abs(left[1] - position[str(i)][1])
            r_distance = abs(right[0] - position[str(i)][0]) + abs(right[1] - position[str(i)][1])
            if l_distance < r_distance:
                answer += 'L'
                left = position[str(i)]

            elif l_distance > r_distance:
                answer += 'R'
                right = position[str(i)]
            else:
                if hand == 'left':
                    answer += 'L'
                    left = position[str(i)]
                else:
                    answer += 'R'
                    right = position[str(i)]

    return answer